File: typescript-sdk/src/opaque.ts
/**
 * OPAQUE password-based key recovery for reeeductio.
 *
 * OPAQUE is an asymmetric PAKE (Password Authenticated Key Exchange) protocol
 * that enables password-based login without exposing passwords or derived keys
 * to the server.
 *
 * Key design points:
 * - OPAQUE is for key recovery only, not authentication
 * - Ed25519 keypairs are randomly generated, not derived from passwords
 * - Password changes re-wrap the same keypair; public keys never change
 * - Credentials are wrapped using OPAQUE's export_key with HKDF + AES-GCM
 */

import * as opaque from '@serenity-kit/opaque';
import {
  deriveKey,
  encryptAesGcm,
  decryptAesGcm,
  encodeBase64,
  decodeBase64,
  concatBytes,
  stringToBytes,
  toUserId,
  generateKeyPair,
} from './crypto.js';
import { createApiError, OpaqueError } from './exceptions.js';
import type {
  ApiError,
  KeyPair,
  OpaqueRegisterInitResponse,
  OpaqueRegisterFinishResponse,
  OpaqueLoginInitResponse,
  OpaqueLoginFinishResponse,
  OpaqueUserRecord,
  OpaqueCredentials,
  OpaqueRegistrationResult,
} from './types.js';

/** Info string for HKDF when deriving credential wrapping key */
const CREDENTIAL_WRAP_INFO = 'reeeductio-credential-wrap';

// ============================================================
// Credential Wrapping/Unwrapping
// ============================================================

/**
 * Wrap credentials (privateKey + symmetricRoot) using OPAQUE's export_key.
 *
 * @param exportKey - 64-byte export key from OPAQUE registration/login
 * @param privateKey - 32-byte Ed25519 private key
 * @param symmetricRoot - 32-byte symmetric root key
 * @returns Encrypted credentials blob (nonce + ciphertext + tag)
 */
export function wrapCredentials(
  exportKey: Uint8Array,
  privateKey: Uint8Array,
  symmetricRoot: Uint8Array
): Uint8Array {
  if (privateKey.length !== 32) {
    throw new OpaqueError(`Private key must be 32 bytes, got ${privateKey.length}`);
  }
  if (symmetricRoot.length !== 32) {
    throw new OpaqueError(`Symmetric root must be 32 bytes, got ${symmetricRoot.length}`);
  }

  // Derive wrapping key from export_key
  const wrapKey = deriveKey(exportKey, CREDENTIAL_WRAP_INFO);

  // Concatenate privateKey and symmetricRoot (64 bytes total)
  const plaintext = concatBytes(privateKey, symmetricRoot);

  // Encrypt with AES-256-GCM
  return encryptAesGcm(plaintext, wrapKey);
}

/**
 * Unwrap credentials using OPAQUE's export_key.
 *
 * @param exportKey - 64-byte export key from OPAQUE login
 * @param encryptedCredentials - Encrypted credentials blob
 * @returns Object with privateKey and symmetricRoot
 */
export function unwrapCredentials(
  exportKey: Uint8Array,
  encryptedCredentials: Uint8Array
): { privateKey: Uint8Array; symmetricRoot: Uint8Array } {
  // Derive wrapping key from export_key
  const wrapKey = deriveKey(exportKey, CREDENTIAL_WRAP_INFO);

  // Decrypt
  const plaintext = decryptAesGcm(encryptedCredentials, wrapKey);

  if (plaintext.length !== 64) {
    throw new OpaqueError(`Decrypted credentials must be 64 bytes, got ${plaintext.length}`);
  }

  return {
    privateKey: plaintext.slice(0, 32),
    symmetricRoot: plaintext.slice(32, 64),
  };
}

// ============================================================
// Low-level API Functions
// ============================================================

/**
 * Start OPAQUE registration (step 1 of 2).
 *
 * @param fetchFn - Fetch implementation
 * @param baseUrl - API base URL
 * @param token - JWT bearer token (required for registration)
 * @param spaceId - Space identifier
 * @param username - Username to register
 * @param registrationRequest - OPAQUE RegistrationRequest (base64)
 * @returns OPAQUE RegistrationResponse (base64)
 */
export async function opaqueRegisterInit(
  fetchFn: typeof fetch,
  baseUrl: string,
  token: string,
  spaceId: string,
  username: string,
  registrationRequest: string
): Promise<OpaqueRegisterInitResponse> {
  const url = `${baseUrl}/spaces/${spaceId}/opaque/register/init`;

  const response = await fetchFn(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`,
    },
    body: JSON.stringify({
      username,
      registration_request: registrationRequest,
    }),
  });

  if (!response.ok) {
    let error: ApiError | undefined;
    try {
      error = (await response.json()) as ApiError;
    } catch {
      // Ignore JSON parse errors
    }
    throw createApiError(response.status, error);
  }

  return response.json() as Promise<OpaqueRegisterInitResponse>;
}

/**
 * Complete OPAQUE registration (step 2 of 2).
 *
 * @param fetchFn - Fetch implementation
 * @param baseUrl - API base URL
 * @param token - JWT bearer token (required for registration)
 * @param spaceId - Space identifier
 * @param username - Username (must match init)
 * @param registrationRecord - OPAQUE RegistrationRecord (base64)
 * @returns OPAQUE PasswordFile (base64)
 */
export async function opaqueRegisterFinish(
  fetchFn: typeof fetch,
  baseUrl: string,
  token: string,
  spaceId: string,
  username: string,
  registrationRecord: string
): Promise<OpaqueRegisterFinishResponse> {
  const url = `${baseUrl}/spaces/${spaceId}/opaque/register/finish`;

  const response = await fetchFn(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`,
    },
    body: JSON.stringify({
      username,
      registration_record: registrationRecord,
    }),
  });

  if (!response.ok) {
    let error: ApiError | undefined;
    try {
      error = (await response.json()) as ApiError;
    } catch {
      // Ignore JSON parse errors
    }
    throw createApiError(response.status, error);
  }

  return response.json() as Promise<OpaqueRegisterFinishResponse>;
}

/**
 * Start OPAQUE login (step 1 of 2).
 *
 * @param fetchFn - Fetch implementation
 * @param baseUrl - API base URL
 * @param spaceId - Space identifier
 * @param username - Username to log in
 * @param credentialRequest - OPAQUE CredentialRequest (base64)
 * @returns OPAQUE CredentialResponse (base64)
 */
export async function opaqueLoginInit(
  fetchFn: typeof fetch,
  baseUrl: string,
  spaceId: string,
  username: string,
  credentialRequest: string
): Promise<OpaqueLoginInitResponse> {
  const url = `${baseUrl}/spaces/${spaceId}/opaque/login/init`;

  const response = await fetchFn(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      username,
      credential_request: credentialRequest,
    }),
  });

  if (!response.ok) {
    let error: ApiError | undefined;
    try {
      error = (await response.json()) as ApiError;
    } catch {
      // Ignore JSON parse errors
    }
    throw createApiError(response.status, error);
  }

  return response.json() as Promise<OpaqueLoginInitResponse>;
}

/**
 * Complete OPAQUE login (step 2 of 2).
 *
 * @param fetchFn - Fetch implementation
 * @param baseUrl - API base URL
 * @param spaceId - Space identifier
 * @param username - Username (must match init)
 * @param credentialFinalization - OPAQUE CredentialFinalization (base64)
 * @returns Encrypted credentials and public key
 */
export async function opaqueLoginFinish(
  fetchFn: typeof fetch,
  baseUrl: string,
  spaceId: string,
  username: string,
  credentialFinalization: string
): Promise<OpaqueLoginFinishResponse> {
  const url = `${baseUrl}/spaces/${spaceId}/opaque/login/finish`;

  const response = await fetchFn(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      username,
      credential_finalization: credentialFinalization,
    }),
  });

  if (!response.ok) {
    let error: ApiError | undefined;
    try {
      error = (await response.json()) as ApiError;
    } catch {
      // Ignore JSON parse errors
    }
    throw createApiError(response.status, error);
  }

  return response.json() as Promise<OpaqueLoginFinishResponse>;
}

// ============================================================
// High-level Registration Flow
// ============================================================

/**
 * Perform full OPAQUE registration for a user.
 *
 * This function:
 * 1. Generates a new Ed25519 keypair (or uses provided one)
 * 2. Performs the OPAQUE registration protocol
 * 3. Wraps credentials with the export key
 * 4. Stores the OPAQUE record via the /data API
 *
 * @param options - Registration options
 * @returns Registration result with username and public key
 */
export async function performOpaqueRegistration(options: {
  fetchFn: typeof fetch;
  baseUrl: string;
  token: string;
  spaceId: string;
  username: string;
  password: string;
  /** Optional keypair to register (generates new one if not provided) */
  keyPair?: KeyPair;
  /** The 32-byte symmetric root for this space */
  symmetricRoot: Uint8Array;
  /** Signing key for storing the OPAQUE record */
  signingPrivateKey: Uint8Array;
  /** Public key of the signer (for signed_by field) */
  signerId: string;
}): Promise<OpaqueRegistrationResult> {
  const {
    fetchFn,
    baseUrl,
    token,
    spaceId,
    username,
    password,
    symmetricRoot,
    signingPrivateKey,
    signerId,
  } = options;

  // Generate keypair if not provided
  const keyPair = options.keyPair ?? await generateKeyPair();
  const publicKey = toUserId(keyPair.publicKey);

  // Step 1: Create client registration state
  const clientRegistration = opaque.client.startRegistration({ password });

  // Step 2: Send registration request to server
  const initResponse = await opaqueRegisterInit(
    fetchFn,
    baseUrl,
    token,
    spaceId,
    username,
    clientRegistration.registrationRequest
  );

  // Step 3: Finish registration on client side
  const registrationResult = opaque.client.finishRegistration({
    password,
    registrationResponse: initResponse.registration_response,
    clientRegistrationState: clientRegistration.clientRegistrationState,
  });

  // Step 4: Send registration record to server
  const finishResponse = await opaqueRegisterFinish(
    fetchFn,
    baseUrl,
    token,
    spaceId,
    username,
    registrationResult.registrationRecord
  );

  // Step 5: Wrap credentials with export key
  const exportKey = decodeBase64(registrationResult.exportKey);
  const encryptedCredentials = wrapCredentials(
    exportKey,
    keyPair.privateKey,
    symmetricRoot
  );

  // Step 6: Create OPAQUE user record
  const record: OpaqueUserRecord = {
    password_file: finishResponse.password_file,
    encrypted_credentials: encodeBase64(encryptedCredentials),
    public_key: publicKey,
  };

  // Step 7: Store via /data API (signed by the registering user)
  const recordJson = JSON.stringify(record);
  const recordData = stringToBytes(recordJson);

  // Import setData function dynamically to avoid circular dependency
  const { setData } = await import('./kvdata.js');

  await setData(
    fetchFn,
    baseUrl,
    token,
    spaceId,
    `opaque/users/${username}`,
    recordData,
    signerId,
    signingPrivateKey
  );

  return {
    username,
    publicKey,
  };
}

// ============================================================
// High-level Login Flow
// ============================================================

/**
 * Perform full OPAQUE login to recover credentials.
 *
 * This function:
 * 1. Performs the OPAQUE login protocol
 * 2. Unwraps credentials with the export key
 * 3. Returns the recovered keypair and symmetric root
 *
 * After calling this, use the recovered credentials to authenticate
 * via the standard Ed25519 challenge-response flow.
 *
 * @param options - Login options
 * @returns Recovered credentials (privateKey, symmetricRoot, publicKey)
 */
export async function performOpaqueLogin(options: {
  fetchFn: typeof fetch;
  baseUrl: string;
  spaceId: string;
  username: string;
  password: string;
}): Promise<OpaqueCredentials> {
  const { fetchFn, baseUrl, spaceId, username, password } = options;

  // Step 1: Create client login state
  const clientLogin = opaque.client.startLogin({ password });

  // Step 2: Send credential request to server
  // Note: The library uses 'startLoginRequest' but the API expects 'credential_request'
  const initResponse = await opaqueLoginInit(
    fetchFn,
    baseUrl,
    spaceId,
    username,
    clientLogin.startLoginRequest
  );

  // Step 3: Finish login on client side
  // Note: The library uses 'loginResponse' but the API returns 'credential_response'
  const loginResult = opaque.client.finishLogin({
    password,
    loginResponse: initResponse.credential_response,
    clientLoginState: clientLogin.clientLoginState,
  });

  if (!loginResult) {
    throw new OpaqueError('OPAQUE login failed: invalid password or protocol error');
  }

  // Step 4: Send credential finalization to server
  // Note: The library uses 'finishLoginRequest' but the API expects 'credential_finalization'
  const finishResponse = await opaqueLoginFinish(
    fetchFn,
    baseUrl,
    spaceId,
    username,
    loginResult.finishLoginRequest
  );

  // Step 5: Unwrap credentials with export key
  const exportKey = decodeBase64(loginResult.exportKey);
  const encryptedCredentials = decodeBase64(finishResponse.encrypted_credentials);

  const { privateKey, symmetricRoot } = unwrapCredentials(exportKey, encryptedCredentials);

  return {
    privateKey,
    symmetricRoot,
    publicKey: finishResponse.public_key,
  };
}

/**
 * Login with OPAQUE and create a Space client.
 *
 * This is a convenience function that performs OPAQUE login, recovers
 * credentials, and returns a fully initialized Space client.
 *
 * @param options - Login options
 * @returns A Space client authenticated with the recovered credentials
 */
export async function loginWithOpaque(options: {
  baseUrl: string;
  spaceId: string;
  username: string;
  password: string;
  fetch?: typeof fetch;
}): Promise<{
  credentials: OpaqueCredentials;
  keyPair: KeyPair;
}> {
  const fetchFn = options.fetch ?? fetch;

  const credentials = await performOpaqueLogin({
    fetchFn,
    baseUrl: options.baseUrl,
    spaceId: options.spaceId,
    username: options.username,
    password: options.password,
  });

  // Reconstruct keypair from recovered private key
  // The public key is verified against the one returned by the server
  const { extractPublicKey } = await import('./crypto.js');
  const expectedPublicKey = extractPublicKey(credentials.publicKey);

  // Derive public key from private key to verify
  const ed = await import('@noble/ed25519');
  const derivedPublicKey = await ed.getPublicKeyAsync(credentials.privateKey);

  // Verify the public key matches
  if (derivedPublicKey.length !== expectedPublicKey.length ||
      !derivedPublicKey.every((v, i) => v === expectedPublicKey[i])) {
    throw new OpaqueError('Public key mismatch: recovered credentials do not match server record');
  }

  const keyPair: KeyPair = {
    privateKey: credentials.privateKey,
    publicKey: derivedPublicKey,
  };

  return { credentials, keyPair };
}

File: python-sdk/reeeductio/opaque.py
"""
OPAQUE protocol client implementation for reeeductio.

Provides password-based key recovery using OPAQUE (Oblivious Pseudo-Random Function
+ Authenticated Key Exchange). OPAQUE enables password-based login without exposing
passwords or derived keys to the server.

Key design points:
- OPAQUE is for key recovery only, not authentication
- Ed25519 keypairs are randomly generated, not derived from passwords
- Credentials are wrapped using OPAQUE's export_key with HKDF + AES-GCM
- After OPAQUE login, the client must still authenticate via Ed25519 challenge-response
"""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
import httpx

from .crypto import Ed25519KeyPair, decode_base64, derive_key, encrypt_aes_gcm, decrypt_aes_gcm, encode_base64
from opaque_snake import (
    OpaqueClient,
    RegistrationRequest,
    RegistrationResponse,
    CredentialRequest,
    CredentialResponse,
    CredentialFinalization,
    ClientRegistrationState,
    ClientLoginState,
)

from .exceptions import (
    AuthenticationError,
    OpaqueError,
    OpaqueNotEnabledError,
    ValidationError,
)

# HKDF info string for credential wrapping (must match backend)
CREDENTIAL_WRAP_INFO = "reeeductio-credential-wrap"


@dataclass
class OpaqueCredentials:
    """
    Credentials recovered from OPAQUE login.

    Attributes:
        keypair: Ed25519 key pair for authentication and signing
        symmetric_root: 256-bit root key for HKDF derivation
        public_key: User's public key identifier (44-char base64)
    """
    keypair: Ed25519KeyPair
    symmetric_root: bytes
    public_key: str


def wrap_credentials(export_key: bytes, private_key: bytes, symmetric_root: bytes) -> bytes:
    """
    Wrap credentials using a key derived from OPAQUE's export_key.

    Args:
        export_key: OPAQUE export_key from registration/login
        private_key: 32-byte Ed25519 private key
        symmetric_root: 32-byte symmetric root key

    Returns:
        AES-GCM encrypted credentials (nonce + ciphertext + tag)
    """
    # Derive wrapping key from export_key
    wrap_key = derive_key(export_key, CREDENTIAL_WRAP_INFO)

    # Concatenate credentials: privateKey (32) || symmetricRoot (32) = 64 bytes
    plaintext = private_key + symmetric_root

    # Encrypt with AES-GCM
    return encrypt_aes_gcm(plaintext, wrap_key)


def unwrap_credentials(export_key: bytes, encrypted_credentials: bytes) -> tuple[bytes, bytes]:
    """
    Unwrap credentials using a key derived from OPAQUE's export_key.

    Args:
        export_key: OPAQUE export_key from login
        encrypted_credentials: AES-GCM encrypted credentials

    Returns:
        Tuple of (private_key, symmetric_root)

    Raises:
        ValueError: If decryption fails
    """
    # Derive wrapping key from export_key
    wrap_key = derive_key(export_key, CREDENTIAL_WRAP_INFO)

    # Decrypt
    plaintext = decrypt_aes_gcm(encrypted_credentials, wrap_key)

    # Split: privateKey (32) || symmetricRoot (32)
    if len(plaintext) != 64:
        raise ValueError(f"Invalid credential length: expected 64 bytes, got {len(plaintext)}")

    private_key = plaintext[:32]
    symmetric_root = plaintext[32:]

    return private_key, symmetric_root


def opaque_login(
    base_url: str,
    space_id: str,
    username: str,
    password: str,
) -> OpaqueCredentials:
    """
    Perform OPAQUE login to recover credentials from a password.

    This is a standalone function that does not require authentication.
    After OPAQUE login, use the returned credentials to authenticate
    via the standard Ed25519 challenge-response flow.

    Args:
        base_url: Base URL of the reeeductio server
        space_id: Space identifier
        username: OPAQUE username
        password: User's password

    Returns:
        OpaqueCredentials containing keypair, symmetric_root, and public_key

    Raises:
        OpaqueNotEnabledError: If OPAQUE is not enabled for this space
        OpaqueError: If OPAQUE protocol fails
        AuthenticationError: If password is incorrect

    Example:
        credentials = opaque_login(
            base_url="http://localhost:8000",
            space_id="C...",
            username="alice",
            password="secret123"
        )

        # Use recovered credentials to create a Space client
        space = Space(
            space_id=space_id,
            member_id=credentials.keypair.to_user_id(),
            private_key=credentials.keypair.private_key,
            symmetric_root=credentials.symmetric_root,
            base_url=base_url,
        )
    """

    with httpx.Client(base_url=base_url) as client:
        # Step 1: Create client-side OPAQUE state and credential request
        opaque_client = OpaqueClient()
        credential_request, client_login_state = opaque_client.start_login(password)

        # Step 2: Send credential request to server
        try:
            response = client.post(
                f"/spaces/{space_id}/opaque/login/init",
                json={
                    "username": username,
                    "credential_request": base64.b64encode(credential_request.to_bytes()).decode("ascii"),
                },
            )
            response.raise_for_status()
            init_data = response.json()
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 501:
                raise OpaqueNotEnabledError(f"OPAQUE is not enabled for space {space_id}") from e
            elif e.response.status_code == 401:
                raise AuthenticationError(f"User '{username}' not found") from e
            elif e.response.status_code == 429:
                raise OpaqueError(f"Too many login attempts: {e.response.text}") from e
            raise OpaqueError(f"OPAQUE login/init failed: {e.response.text}") from e

        # Step 3: Process server response to get export_key and finalization
        try:
            credential_response_bytes = base64.b64decode(init_data["credential_response"])
            credential_response = CredentialResponse.from_bytes(credential_response_bytes)
        except Exception as e:
            raise OpaqueError(f"Invalid credential response: {e}") from e

        try:
            login_result = opaque_client.finish_login(
                credential_response, client_login_state, password
            )
            credential_finalization = login_result.finalization
            export_key = login_result.session_keys.export_key
        except Exception as e:
            raise AuthenticationError(f"OPAQUE authentication failed: {e}") from e

        # Step 4: Send finalization to server and get encrypted credentials
        try:
            response = client.post(
                f"/spaces/{space_id}/opaque/login/finish",
                json={
                    "username": username,
                    "credential_finalization": base64.b64encode(credential_finalization.to_bytes()).decode("ascii"),
                },
            )
            response.raise_for_status()
            finish_data = response.json()
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 401:
                raise AuthenticationError("OPAQUE authentication failed: invalid password") from e
            elif e.response.status_code == 404:
                raise OpaqueError(f"Login session expired: {e.response.text}") from e
            elif e.response.status_code == 429:
                raise OpaqueError(f"Too many failed attempts: {e.response.text}") from e
            raise OpaqueError(f"OPAQUE login/finish failed: {e.response.text}") from e

        # Step 5: Unwrap credentials using export_key
        encrypted_credentials = base64.b64decode(finish_data["encrypted_credentials"])
        public_key = finish_data["public_key"]

        try:
            private_key, symmetric_root = unwrap_credentials(export_key, encrypted_credentials)
        except Exception as e:
            raise OpaqueError(f"Failed to unwrap credentials: {e}") from e

        # Extract raw public key from typed identifier for keypair
        from .crypto import Ed25519KeyPair
        raw_public_key = Ed25519KeyPair.from_typed_public_key(public_key)

        return OpaqueCredentials(
            keypair=Ed25519KeyPair(private_key=private_key, public_key=raw_public_key),
            symmetric_root=symmetric_root,
            public_key=public_key,
        )


async def opaque_login_async(
    base_url: str,
    space_id: str,
    username: str,
    password: str,
) -> OpaqueCredentials:
    """
    Async version of opaque_login.

    See opaque_login for full documentation.
    """

    async with httpx.AsyncClient(base_url=base_url) as client:
        # Step 1: Create client-side OPAQUE state and credential request
        opaque_client = OpaqueClient()
        credential_request, client_login_state = opaque_client.start_login(password)

        # Step 2: Send credential request to server
        try:
            response = await client.post(
                f"/spaces/{space_id}/opaque/login/init",
                json={
                    "username": username,
                    "credential_request": base64.b64encode(credential_request.to_bytes()).decode("ascii"),
                },
            )
            response.raise_for_status()
            init_data = response.json()
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 501:
                raise OpaqueNotEnabledError(f"OPAQUE is not enabled for space {space_id}") from e
            elif e.response.status_code == 401:
                raise AuthenticationError(f"User '{username}' not found") from e
            elif e.response.status_code == 429:
                raise OpaqueError(f"Too many login attempts: {e.response.text}") from e
            raise OpaqueError(f"OPAQUE login/init failed: {e.response.text}") from e

        # Step 3: Process server response
        try:
            credential_response_bytes = base64.b64decode(init_data["credential_response"])
            credential_response = CredentialResponse.from_bytes(credential_response_bytes)
        except Exception as e:
            raise OpaqueError(f"Invalid credential response: {e}") from e

        try:
            login_result = opaque_client.finish_login(
                credential_response, client_login_state, password
            )
            credential_finalization = login_result.finalization
            export_key = login_result.session_keys.export_key
        except Exception as e:
            raise AuthenticationError(f"OPAQUE authentication failed: {e}") from e

        # Step 4: Send finalization
        try:
            response = await client.post(
                f"/spaces/{space_id}/opaque/login/finish",
                json={
                    "username": username,
                    "credential_finalization": base64.b64encode(credential_finalization.to_bytes()).decode("ascii"),
                },
            )
            response.raise_for_status()
            finish_data = response.json()
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 401:
                raise AuthenticationError("OPAQUE authentication failed: invalid password") from e
            elif e.response.status_code == 404:
                raise OpaqueError(f"Login session expired: {e.response.text}") from e
            elif e.response.status_code == 429:
                raise OpaqueError(f"Too many failed attempts: {e.response.text}") from e
            raise OpaqueError(f"OPAQUE login/finish failed: {e.response.text}") from e

        # Step 5: Unwrap credentials
        encrypted_credentials = base64.b64decode(finish_data["encrypted_credentials"])
        public_key = finish_data["public_key"]

        try:
            private_key, symmetric_root = unwrap_credentials(export_key, encrypted_credentials)
        except Exception as e:
            raise OpaqueError(f"Failed to unwrap credentials: {e}") from e

        from .crypto import Ed25519KeyPair
        raw_public_key = Ed25519KeyPair.from_typed_public_key(public_key)

        return OpaqueCredentials(
            keypair=Ed25519KeyPair(private_key=private_key, public_key=raw_public_key),
            symmetric_root=symmetric_root,
            public_key=public_key,
        )


def opaque_register(
    client: httpx.Client,
    space_id: str,
    username: str,
    password: str,
    user_id: str,
    private_key: bytes,
    symmetric_root: bytes,
) -> str:
    """
    Register OPAQUE credentials for password-based login.

    This function requires authentication - the caller must have a valid JWT token
    in the provided httpx.Client. Typically called from Space.opaque_register().

    The registration flow:
    1. Verify user_id encodes the correct public key for the private_key
    2. Send registration request to server
    3. Process server response to get export_key and registration record
    4. Server returns password_file
    5. Wrap credentials with export_key
    6. Store complete record via /data API

    Args:
        client: Authenticated httpx.Client
        space_id: Space identifier
        username: OPAQUE username to register
        password: Password for future logins
        user_id: Typed identifier string (USER or TOOL) for the public key
        private_key: 32-byte Ed25519 private key matching user_id
        symmetric_root: Symmetric root key to wrap

    Returns:
        The username that was registered

    Raises:
        OpaqueNotEnabledError: If OPAQUE is not enabled for this space
        OpaqueError: If registration fails
        ValidationError: If username already exists or user_id doesn't match private_key
    """

    # Verify user_id matches private_key
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
    from cryptography.hazmat.primitives import serialization

    try:
        expected_public_key = Ed25519KeyPair.from_typed_public_key(user_id)
    except ValueError as e:
        raise ValidationError(f"Invalid user_id: {e}") from e

    try:
        private_key_obj = Ed25519PrivateKey.from_private_bytes(private_key)
        derived_public_key = private_key_obj.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )
    except Exception as e:
        raise ValidationError(f"Invalid private_key: {e}") from e

    if derived_public_key != expected_public_key:
        raise ValidationError("user_id does not match the provided private_key")

    # Step 1: Create client-side registration request
    opaque_client = OpaqueClient()
    registration_request, client_reg_state = opaque_client.start_registration(password)

    # Step 2: Send registration request to server
    try:
        response = client.post(
            f"/spaces/{space_id}/opaque/register/init",
            json={
                "username": username,
                "registration_request": base64.b64encode(registration_request.to_bytes()).decode("ascii"),
            },
        )
        response.raise_for_status()
        init_data = response.json()
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 501:
            raise OpaqueNotEnabledError(f"OPAQUE is not enabled for space {space_id}") from e
        elif e.response.status_code == 409:
            raise ValidationError(f"Username '{username}' is already registered") from e
        elif e.response.status_code == 401:
            raise AuthenticationError(f"Authentication required for OPAQUE registration") from e
        raise OpaqueError(f"OPAQUE register/init failed: {e.response.text}") from e

    # Step 3: Process server response
    try:
        registration_response_bytes = base64.b64decode(init_data["registration_response"])
        registration_response = RegistrationResponse.from_bytes(registration_response_bytes)
    except Exception as e:
        raise OpaqueError(f"Invalid registration response: {e}") from e

    # Step 4: Finish client-side registration to get export_key and registration record
    try:
        reg_result = opaque_client.finish_registration(
            registration_response, client_reg_state, password
        )
        registration_record = reg_result.upload
        export_key = reg_result.export_key
    except Exception as e:
        raise OpaqueError(f"Failed to complete registration: {e}") from e

    # Step 5: Send registration record to server to get password_file
    try:
        response = client.post(
            f"/spaces/{space_id}/opaque/register/finish",
            json={
                "username": username,
                "registration_record": base64.b64encode(registration_record.to_bytes()).decode("ascii"),
            },
        )
        response.raise_for_status()
        finish_data = response.json()
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 401:
            raise AuthenticationError(f"Authentication failed or session mismatch") from e
        elif e.response.status_code == 404:
            raise OpaqueError(f"Registration session expired") from e
        elif e.response.status_code == 409:
            raise ValidationError(f"Username '{username}' is already registered") from e
        raise OpaqueError(f"OPAQUE register/finish failed: {e.response.text}") from e

    password_file = finish_data["password_file"]

    # Step 7: Wrap credentials with export_key
    encrypted_credentials = wrap_credentials(export_key, private_key, symmetric_root)
    encrypted_credentials_b64 = base64.b64encode(encrypted_credentials).decode("ascii")

    # Step 8: Assemble complete OPAQUE record
    opaque_record = {
        "password_file": password_file,
        "encrypted_credentials": encrypted_credentials_b64,
        "public_key": user_id,
    }

    # Step 9: Store via /data API
    from . import kvdata

    kvdata.set_data(
        client=client,
        space_id=space_id,
        path=f"opaque/users/{username}",
        data=json.dumps(opaque_record).encode("utf-8"),
        signed_by=user_id,
        private_key=private_key,
    )

    return username


async def opaque_register_async(
    client: httpx.AsyncClient,
    space_id: str,
    username: str,
    password: str,
    user_id: str,
    private_key: bytes,
    symmetric_root: bytes,
) -> str:
    """
    Async version of opaque_register.

    See opaque_register for full documentation.
    """

    # Verify user_id matches private_key
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
    from cryptography.hazmat.primitives import serialization

    try:
        expected_public_key = Ed25519KeyPair.from_typed_public_key(user_id)
    except ValueError as e:
        raise ValidationError(f"Invalid user_id: {e}") from e

    try:
        private_key_obj = Ed25519PrivateKey.from_private_bytes(private_key)
        derived_public_key = private_key_obj.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )
    except Exception as e:
        raise ValidationError(f"Invalid private_key: {e}") from e

    if derived_public_key != expected_public_key:
        raise ValidationError("user_id does not match the provided private_key")

    # Step 1: Create client-side registration request
    opaque_client = OpaqueClient()
    registration_request, client_reg_state = opaque_client.start_registration(password)

    # Step 2: Send registration request to server
    try:
        response = await client.post(
            f"/spaces/{space_id}/opaque/register/init",
            json={
                "username": username,
                "registration_request": base64.b64encode(registration_request.to_bytes()).decode("ascii"),
            },
        )
        response.raise_for_status()
        init_data = response.json()
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 501:
            raise OpaqueNotEnabledError(f"OPAQUE is not enabled for space {space_id}") from e
        elif e.response.status_code == 409:
            raise ValidationError(f"Username '{username}' is already registered") from e
        elif e.response.status_code == 401:
            raise AuthenticationError(f"Authentication required for OPAQUE registration") from e
        raise OpaqueError(f"OPAQUE register/init failed: {e.response.text}") from e

    # Step 3: Process server response
    try:
        registration_response_bytes = base64.b64decode(init_data["registration_response"])
        registration_response = RegistrationResponse.from_bytes(registration_response_bytes)
    except Exception as e:
        raise OpaqueError(f"Invalid registration response: {e}") from e

    # Step 4: Finish client-side registration
    try:
        reg_result = opaque_client.finish_registration(
            registration_response, client_reg_state, password
        )
        registration_record = reg_result.upload
        export_key = reg_result.export_key
    except Exception as e:
        raise OpaqueError(f"Failed to complete registration: {e}") from e

    # Step 5: Send registration record to server
    try:
        response = await client.post(
            f"/spaces/{space_id}/opaque/register/finish",
            json={
                "username": username,
                "registration_record": base64.b64encode(registration_record.to_bytes()).decode("ascii"),
            },
        )
        response.raise_for_status()
        finish_data = response.json()
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 401:
            raise AuthenticationError(f"Authentication failed or session mismatch") from e
        elif e.response.status_code == 404:
            raise OpaqueError(f"Registration session expired") from e
        elif e.response.status_code == 409:
            raise ValidationError(f"Username '{username}' is already registered") from e
        raise OpaqueError(f"OPAQUE register/finish failed: {e.response.text}") from e

    password_file = finish_data["password_file"]

    # Step 7: Wrap credentials
    encrypted_credentials = wrap_credentials(export_key, private_key, symmetric_root)
    encrypted_credentials_b64 = base64.b64encode(encrypted_credentials).decode("ascii")

    # Step 8: Assemble record
    opaque_record = {
        "password_file": password_file,
        "encrypted_credentials": encrypted_credentials_b64,
        "public_key": user_id,
    }

    # Step 9: Store via /data API
    from . import kvdata

    await kvdata.set_data_async(
        client=client,
        space_id=space_id,
        path=f"opaque/users/{username}",
        data=json.dumps(opaque_record).encode("utf-8"),
        signed_by=user_id,
        private_key=private_key,
    )

    return username


# Constants for OPAQUE setup
OPAQUE_SERVER_SETUP_PATH = "opaque/server/setup"
OPAQUE_USER_ROLE_ID = "opaque-user"
OPAQUE_USER_CAP_ID = "cap_create_opaque_user"
